test/modules/load-gen/PriorityMessageSender.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace LoadGen
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Extensions.Logging;
+
+    class PriorityMessageSender : SenderBase
+    {
+        readonly string[] outputs = new string[] { "pri0", "pri1", "pri2", "pri3" };
+
+        readonly Random rng = new Random();
+
+        public PriorityMessageSender(
+            ILogger logger,
+            ModuleClient moduleClient,
+            Guid batchId,
+            string trackingId)
+            : base(logger, moduleClient, batchId, trackingId)
+        {
+        }
+
+        public async override Task RunAsync(CancellationTokenSource cts, DateTime testStartAt)
+        {
+            bool firstMessageWhileOffline = true;
+            var priorityAndSequenceList = new List<(int, long)>();
+            long messageIdCounter = 1;
+            while (!cts.IsCancellationRequested &&
+                (Settings.Current.TestDuration == TimeSpan.Zero || DateTime.UtcNow - testStartAt < Settings.Current.TestDuration))
+            {
+                try
+                {
+                    int priority = this.rng.Next(4);
+                    string output = this.outputs[priority];
+
+                    await this.SendEventAsync(messageIdCounter, output);
+
+                    // We need to set the first message because of the way priority queue logic works
```

Comment:
@richma-ms we didn't mention this behavior in design spec.  How difficult to change this behavior (always retyr first message sent while offline)?

Reply:
the behavior is not always retry the first message, but rather whichever message loop happens to be sending when connectivity went down.  Due to the way this test is setup, it ends up being the first message because connectivity is already down by the time the first message gets enqueued.

Changing this is very difficult, as it'd involve fundamentally changing how retry and message checkpointing works.  We can't promise 100% accurate priority order anyways, even if we reworked retry, there will still be race conditions that exists depending on where code execution is at when connectivity gets cut/resumed.  So what we have now is best effort for the least development cost, and as far as we know it's good enough for the customer scenarios that we have.

---
In reply to: [397497250](https://github.com/Azure/iotedge/pull/2735#discussion_r397497250) [](ancestors = 397497250)